segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/DebugLogWrapper.java

```diff
@@ -128,6 +152,98 @@ public void disable() throws DurableDataLogException {
         this.log.disable();
     }
 
+    /**
+     * Performs a {@link BookKeeperLog}-{@link LedgerHandle} reconciliation for this {@link BookKeeperLog} subject to the
+     * following rules:
+     * - Any {@link LedgerHandle}s that list this {@link BookKeeperLog} as their owner will be added to this {@link BookKeeperLog}'s
+     * list of ledgers (if they're non-empty and haven't been truncated out).
+     * - Any {@link LedgerMetadata} instances in this {@link BookKeeperLog} that point to inexistent {@link LedgerHandle}s
+     * will be removed.
+     *
+     * @param candidateLedgers A List of {@link LedgerHandle}s that contain all the Ledgers that this {@link BookKeeperLog}
+     *                         should contain. This could be the list of all BookKeeper Ledgers or a subset, as long as
+     *                         it contains all Ledgers that list this {@link BookKeeperLog} as their owner.
+     * @return True if something changed (and the metadata is updated), false otherwise.
+     * @throws IllegalStateException   If this BookKeeperLog is not disabled.
+     * @throws DurableDataLogException If an exception occurred while updating the metadata.
+     */
+    public boolean reconcileLedgers(List<LedgerHandle> candidateLedgers) throws DurableDataLogException {
+        // Load metadata and verify if disabled (metadata may be null if it doesn't exist).
+        LogMetadata metadata = this.log.loadMetadata();
+        final long highestLedgerId;
+        if (metadata != null) {
+            Preconditions.checkState(!metadata.isEnabled(), "BookKeeperLog is enabled; cannot reconcile ledgers.");
+            int ledgerCount = metadata.getLedgers().size();
+            if (ledgerCount > 0) {
+                // Get the highest Ledger id from the list of ledgers.
+                highestLedgerId = metadata.getLedgers().get(ledgerCount - 1).getLedgerId();
+            } else if (metadata.getTruncationAddress() != null) {
+                // All Ledgers have been truncated out. Get it from the Truncation Address.
+                highestLedgerId = metadata.getTruncationAddress().getLedgerId();
+            } else {
+                // No information.
+                highestLedgerId = Ledgers.NO_LEDGER_ID;
+            }
+        } else {
+            // No metadata.
+            highestLedgerId = Ledgers.NO_LEDGER_ID;
+        }
+
+        // First, we filter out any Ledger that does not reference this Log as their owner or that are empty.
+        candidateLedgers = candidateLedgers
```

Comment:
What happens with upgrade here?

cc: @RaulGracia

Reply:
Nothing too exciting. This will not work on ledgers that are not tagged, which is why this is important to go in well before any attempt to fix any issues so that each ledger in the system is properly tagged.